serverv2/appmanager/app_manager_full.go

```diff
@@ -0,0 +1,119 @@
+package appmanager
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+)
+
+type Mempool interface {
+}
+
+type AppManagerBuilder struct {
+	InitGenesis map[string]func(ctx context.Context, moduleGenesisBytes []byte) error
+}
+
+func (a *AppManagerBuilder) RegisterInitGenesis(moduleName string, genesisFunc func(ctx context.Context, moduleGenesisBytes []byte) error) {
+	a.InitGenesis[moduleName] = genesisFunc
+}
+
+func (a *AppManagerBuilder) RegisterHandler(moduleName, handlerName string, handler MsgHandler) {
+	panic("...")
+}
+
+type MsgSetKVPairs struct {
+	Pairs []ChangeSet
+}
+
+func (a *AppManagerBuilder) Build() *AppManager {
+	genesis := func(ctx context.Context, genesisBytes []byte) error {
+		genesisMap := map[string][]byte{} // module=> genesis bytes
+		for module, genesisFunc := range a.InitGenesis {
+			err := genesisFunc(ctx, genesisMap[module])
+			if err != nil {
+				return fmt.Errorf("failed to init genesis on module: %s", module)
+			}
+		}
+		return nil
+	}
+	return &AppManager{initGenesis: genesis}
+}
+
+type AppManager struct {
+	// configs
+	checkTxGasLimit uint64
+	queryGasLimit   uint64
+	// configs - end
+
+	db Store
+
+	lastBlockHeight *atomic.Uint64
+
+	initGenesis func(ctx context.Context, genesisBytes []byte) error
+
+	stf *STFAppManager
+
+	mempool Mempool
+}
+
+func (a AppManager) CheckTx(ctx context.Context, txBytes []byte) error {
+	// decode tx
+	tx, err := a.stf.decodeTx(txBytes)
+	if err != nil {
+		return err
+	}
+	// apply validation using last block state
+	checkTxState, err := a.getLatestState(ctx)
+	if err != nil {
+		return err
+	}
+	_, _, err = a.stf.validateTx(ctx, checkTxState, min(a.checkTxGasLimit, tx.GetGasLimit()), tx)
+	if err != nil {
+		return err
+	}
+	// TODO: cache, insert in mempool?
+	return nil
+}
```

Comment:
this would be on the mempool, not sure we need to add this into the appamanger. It doesnt really have anything to do with the app manager unless you are decoding and runing the tx, but in that case app manager runmsg should be passed to mempool instead of mempool being passed here